src/System.Private.ServiceModel/src/System/IdentityModel/Selectors/X509CertificateValidator.cs

```diff
@@ -54,6 +83,77 @@ public override void Validate(X509Certificate2 certificate)
             }
         }
 
+        class PeerTrustValidator : X509CertificateValidator
+        {
+            public override void Validate(X509Certificate2 certificate)
+            {
+                if (certificate == null)
+                    throw DiagnosticUtility.ExceptionUtility.ThrowHelperArgumentNull("certificate");
+
+                Exception exception;
+                if (!TryValidate(certificate, out exception))
+                    throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(exception);
+            }
+
+            static bool StoreContainsCertificate(StoreName storeName, X509Certificate2 certificate)
+            {
+                X509Store store = new X509Store(storeName, StoreLocation.CurrentUser);
+                X509Certificate2Collection certificates = null;
+                try
+                {
+                    store.Open(OpenFlags.ReadOnly);
+                    certificates = store.Certificates.Find(X509FindType.FindByThumbprint, certificate.Thumbprint, false);
+                    return certificates.Count > 0;
+                }
+                finally
+                {
+                    SecurityUtils.ResetAllCertificates(certificates);
+                    store.Dispose();
+                }
+            }
+
+            internal bool TryValidate(X509Certificate2 certificate, out Exception exception)
+            {
+                // Checklist
+                // 1) time validity of cert
+                // 2) in trusted people store
+                // 3) not in disallowed store
+
+                // The following code could be written as:
+                // DateTime now = DateTime.UtcNow;
+                // if (now > certificate.NotAfter.ToUniversalTime() || now < certificate.NotBefore.ToUniversalTime())
```

Comment:
@iamjasonp -- what are the advantages of doing what the comment says if the cert times are local? I wonder if we risk subtle bugs or behavioral changes by diverging from desktop.

Incidentally, if the comments about certs using local time are correct, then our CertificateGenerator may be subtly  doing the wrong thing [here](https://github.com/dotnet/wcf/blob/master/src/System.Private.ServiceModel/tools/CertificateGenerator/CertificateGenerator.cs#L91) and for expired certs [here](https://github.com/dotnet/wcf/blob/master/src/System.Private.ServiceModel/tools/CertificateGenerator/Program.cs#L111).